### PR TITLE
Flexible Hyperparameter Search Space

### PR DESCRIPTION
**Pull request by cjmcgill**
## Description
This PR allows users to choose the hyperparameters to include in the hyperparameter search space. By using the argument `--search_parameter_keywords {list-of-keywords}`, the user can choose the parameters to be included.

Also added informative errors that trigger when checkpoints for a different parameter space are used, to prevent accidents in inappropriate parallelization. Notably, Hyperopt does not raise an error in these situation so implementing usable errors here is important. Also reorganized the hyperparameter optimization section in the readme and moved several code functions over to utils from the main script.

As listed in the readme, these are the available parameters:

Special keywords
* basic - the default set of hyperparameters for search: depth, ffn_num_layers, dropout, and linked_hidden_size.
* linked_hidden_size - search for hidden_size and ffn_hidden_size, but constrained for them to have the same value. If either of the component words are entered in separately, both are searched independently.
* learning_rate - search for max_lr, init_lr, final_lr, and warmup_epochs.
* all - include search for all inidividual keyword options

Individual supported parameters
* activation, aggregation, aggregation_norm, batch_size, depth, dropout, ffn_hidden_size, ffn_num_layers, final_lr, hidden_size, init_lr, max_lr, warmup_epochs

## Example
```
chemprop_hyperopt --data_path <data_path> --dataset_type <type> --num_iters <int> --config_save_path <config_path> --search_parameter_keywords depth activation batch_size
```
## Questions
Too many options can be a bad thing, and in this case can make the search space combinatorially large. Are any of these parameters insensitive enough that it's not worth implementing them? Similarly, are there any important parameters that should be included that aren't? Should we make a strong recommendation about what parameter sets people should use?

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Review by shihchengli — COMMENTED** [review]
Good work on flexible hyperparameter search. The change in this PR is very clear. I left some comments below. I totally agree that we should remove the insensitive hyperparameters to reduce the complexity of search space, but I have no idea which one is possible to be removed. I think the sensitivity of each parameter might vary with different tasks. By the way, should we recommend increasing `--startup_random_iters` (default = 10) when the search space is combinatorially large. Since the hyperparameters are randomly chosen as a warm-up to get an idea of where to start, if there are too many hyperparameters to be tuned, the tunning results may become different each time.

File: chemprop/hyperopt_utils.py [hunk 1]
```diff
@@ -5,11 +5,43 @@
...
+        "depth": hp.quniform("depth", low=2, high=6, q=1),
+        "dropout": hp.quniform("dropout", low=0.0, high=0.4, q=0.05),
+        "ffn_hidden_size": hp.quniform("ffn_hidden_size", low=300, high=2400, q=100),
+        "ffn_num_layers": hp.quniform("ffn_num_layers", low=1, high=3, q=1),
+        "final_lr_ratio": hp.loguniform("final_lr_ratio", low=np.log(1e-4), high=0.),
+        "hidden_size": hp.quniform("hidden_size", low=300, high=2400, q=100),
+        "init_lr_ratio": hp.loguniform("init_lr_ratio", low=np.log(1e-4), high=0.),
+        "linked_hidden_size": hp.quniform("linked_hidden_size", low=300, high=2400, q=100),
+        "max_lr": hp.loguniform("max_lr", low=np.log(1e-6), high=np.log(1e-2)),
+        "warmup_epochs": hp.uniform("warmup_epochs", low=1, high=0.5 * train_epochs)
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:36]: Very good suggestion

File: chemprop/hyperopt_utils.py [hunk 2]
```diff
@@ -158,12 +219,20 @@ def load_manual_trials(manual_trials_dirs: List[str], param_keys: List[str], hyp
...
+        elif 'hidden_size' not in param_keys or 'ffn_hidden_size' not in param_keys:
+            if 'hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'hidden_size') != trial_args['hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument hidden_size than the hyperparameter optimization search trials.')
+            if 'ffn_hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'ffn_hidden_size') != trial_args['ffn_hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument ffn_hidden_size than the hyperparameter optimization search trials.')
+
+        for arg, space_parameter in matching_args:
+            if space_parameter not in param_keys:
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:235]: No, this is intentional. If a parameter is included in the search space, then the manual trial can differ in that parameter and it's okay. We only care about parameters that are not represented in the space parameter keys. And at this step, we're looking at the keywords in the search space, which are a little different than the argument names (`init_lr` versus `init_lr_ratio` for instance).

File: chemprop/hyperparameter_optimization.py [hunk 3]
```diff
@@ -60,21 +59,32 @@ def hyperopt(args: HyperoptArgs) -> None:
...
         # Copy args
         hyper_args = deepcopy(args)
 
         # Update args with hyperparams
         if args.save_dir is not None:
-            folder_name = '_'.join(f'{key}_{value}' for key, value in hyperparams.items())
+            if len(hyperparams) > 5:
+                folder_name = f'trial_seed_{seed}'
+            else:
+                folder_name = '_'.join(f'{key}_{value}' for key, value in hyperparams.items())
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/hyperparameter_optimization.py:74]: We probably should, yes.

**Comment by cjmcgill** [comment]
@shihchengli I just added a change where the default number of random trials is half the total number of trials. I think that will handle it reasonably for when people search many parameters and increase the number of trials accordingly.

File: chemprop/hyperopt_utils.py [hunk 4]
```diff
@@ -5,11 +5,43 @@
...
+        "dropout": hp.quniform("dropout", low=0.0, high=0.4, q=0.05),
+        "ffn_hidden_size": hp.quniform("ffn_hidden_size", low=300, high=2400, q=100),
+        "ffn_num_layers": hp.quniform("ffn_num_layers", low=1, high=3, q=1),
+        "final_lr_ratio": hp.loguniform("final_lr_ratio", low=np.log(1e-4), high=0.),
+        "hidden_size": hp.quniform("hidden_size", low=300, high=2400, q=100),
+        "init_lr_ratio": hp.loguniform("init_lr_ratio", low=np.log(1e-4), high=0.),
+        "linked_hidden_size": hp.quniform("linked_hidden_size", low=300, high=2400, q=100),
+        "max_lr": hp.loguniform("max_lr", low=np.log(1e-6), high=np.log(1e-2)),
+        "warmup_epochs": hp.quniform("warmup_epochs", low=1, high=train_epochs // 2, q=1)
+    }
```
Comment (shihchengli) [inline chemprop/hyperopt_utils.py:37]: Since the searching range of each parameter is hard coded here, users might want to change this range. For example, if `--features_only` is used, only the additional features will be passed to FFN, no graph network, and it doesn't make sense to still search the hyperparameter of `ffn_hidden_size` starting from 300 to 2400. I think it would be better to add some description like "The searching range of each hyperparams has been hard coded, and if you want to change it, please change the object of available_spaces in build_search_space() in  hyperopt_utils.py". I think this change is not necessary to be made since the default setting is generally enough for most cases.

Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:37]: I think this would be useful in the readme and have added it.

File: chemprop/hyperopt_utils.py [hunk 5]
```diff
@@ -158,13 +220,21 @@ def load_manual_trials(manual_trials_dirs: List[str], param_keys: List[str], hyp
...
+        elif 'hidden_size' not in param_keys or 'ffn_hidden_size' not in param_keys:
+            if 'hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'hidden_size') != trial_args['hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument hidden_size than the hyperparameter optimization search trials.')
+            if 'ffn_hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'ffn_hidden_size') != trial_args['ffn_hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument ffn_hidden_size than the hyperparameter optimization search trials.')
+
+        for arg, space_parameter in matching_args:
+            if space_parameter not in param_keys:
```
Comment (shihchengli) [inline chemprop/hyperopt_utils.py:236]: If the `linked_hidden_size` is in `param_keys`, should it be replaced with `hidden_size` and `ffn_hidden_size`? Because if `linked_hidden_size` is specified for hyperparameter optimization, this case we don't want the values of `hidden_size` and `ffn_hidden_size` be the same as the manual trail. However, setting `linked_hidden_size` as an element of `param_keys` cannot reflect this point. Please correct me if I misunderstand this part. Similar questions can be found in `init_lr` and `final_lr`.

File: chemprop/hyperopt_utils.py [hunk 6]
```diff
@@ -116,29 +169,38 @@ def get_hyperopt_seed(seed: int, dir_path: str) -> int:
...
+        ('reaction', None),
+        ('split_type', None),
+        ('warmup_epochs', 'warmup_epochs'),
+        ('split_sizes', None),
+        ('aggregation_norm', 'aggregation_norm'),
+        ('batch_size', 'batch_size'),
+        ('depth', 'depth'),
+        ('dropout', 'dropout'),
+        ('ffn_num_layers', 'ffn_num_layers'),
     ]
```
Comment (shihchengli) [inline chemprop/hyperopt_utils.py:204]: How about adding ('dataset_type', None), ('multiclass_num_classes', None), ('features_generator', None), ('no_features_scaling', None), ('features_only', None) into this list. Not sure if it is necessary to extend this list to check more parameters when manual trails are given.

**Comment by kevingreenman** [comment]
This is a great addition that's been needed for a while. I think the overall interface makes sense, and I like the option of special keywords to select parameters that may be commonly optimized together. Do I understand correctly that the main reason one would want to use `linked_hidden_size` would be to reduce the dimension of the search space by 1, or is there some other reason someone would want these to be the same? Also, should the total number of epochs be a parameter that's included (because that should effect the learning rate schedule from maximum to final)?

File: chemprop/hyperopt_utils.py (hunk 5 above)
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:236]: `param_keys` is the name of the parameter as labeled in the search space. `hyperopt_args` is the source of the model parameters in the new trials if they are not covered in the search space. `trial_args` is the model parameters as they appear in the manual trial. The goal of this block is to determine which parameters are controlled by the search space, and make sure for everything else, the parameter in `hyperopt_args` matches the one in `trial_args`.

`linked_hidden_size`, `hidden_size`, and `ffn_hidden_size` are handled in the code block L223-L233 instead of in the L235 loop. It's special behavior that get's handled there. If it's `linked_hidden_size`, then the hidden sizes can be any value as long as they're the same in the trial args. Then in L227-L233 it does the same check as the L235 loop for the regular hidden sizes, but only following the conditional that `linked_hidden_size` isn't in the search space.

`init_lr` and `final_lr` are still handled by the L235 loop. If you look at the tuple in L187 and L188, the label that's checked against the `param_keys` are the ratio versions.

File: chemprop/hyperopt_utils.py (hunk 6 above)
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:204]: These are good additions, I will add them

**Comment by cjmcgill** [comment]
@kevingreenman yes, the linked allows for a search that covers both with one fewer degrees of freedom. I should note that in the readme.

I considered adding in epochs but didn't because my intuition is that more epochs should always be better (or at least equal). There may be a different optimal lr for a different number of epochs, but for a given lr more epochs should be an improvement. I haven't tested this and could be wrong, but that's the case for a constant lr and intuitively seems like it would be for this lr schedule too. So it didn't seem worthwhile to do a search with it if we already know the answer is as many as practical. But, because lr depends on it, people should preferrably use the number of epochs here that they will eventually use in their model.

File: chemprop/hyperopt_utils.py (hunk 5 above)
Comment (shihchengli) [reply chemprop/hyperopt_utils.py:241]: If `linked_hidden_size` is used, the `hidden_size` and `ffn_hidden_size` would be searched, but L235 will still want to make sure these two parameters in `hyperopt_args` are the same as in `trial_args`. However, these two parameters should have different values between `hyperopt_args` and `trial_args`.

Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:241]: It wouldn't though because the hidden sizes aren't part of the L181 list. So L235 loop wouldn't cover them.

Comment (shihchengli) [reply chemprop/hyperopt_utils.py:241]: You are right! Thanks for all the discussion. I have no further questions about this part. It makes sense to me.

**Comment by shihchengli** [comment]
Everything looks good to me!

File: chemprop/hyperopt_utils.py [hunk 7]
```diff
@@ -18,11 +50,32 @@ def merge_trials(trials: Trials, new_trials_data: List[Dict]) -> Trials:
     :param trials_data: The contents of a hyperopt trials object, `Trials.trials`.
     :return: A hyperopt trials object, merged from the two inputs.
     """
-    max_tid = 0
     if len(trials.trials) > 0:
         max_tid = max([trial['tid'] for trial in trials.trials])
+        trial_keys = list(trials.vals.keys())
+        for trial in trials.trials:
+            new_trial_keys = list(trial['misc']['vals'].keys())
+            if trial_keys != new_trial_keys:
```
Comment (shihchengli) [inline chemprop/hyperopt_utils.py:58]:
```suggestion
            if set(trial_keys) != set(new_trial_keys):
```
The order of keys in `new_trial_keys` and `trial_keys` would be different. The order is not guaranteed in Python 3.5 and earlier but is the same as the insertion order in Python 3.6 and older. It is generally a better way to handle this with set(). Similar problem is in L73.

File: chemprop/hyperparameter_optimization.py [hunk 8]
```diff
@@ -110,6 +117,12 @@ def objective(hyperparams: Dict[str, Union[int, float]], seed: int) -> Dict:
     for i in range(args.num_iters):
         # run fmin and load trials in single steps to allow for parallel operation
         trials = load_trials(dir_path=args.hyperopt_checkpoint_dir, previous_trials=manual_trials)
+        if len(trials) >0 and set(space.keys()) != set(trials.vals.keys()):
```
Comment (shihchengli) [inline chemprop/hyperparameter_optimization.py:120]:
```suggestion
        if len(trials) > 0 and set(space.keys()) != set(trials.vals.keys()):
```

File: chemprop/hyperopt_utils.py (hunk 7 above)
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:58]: Right you are. I made these lists into sets instead, not at the point of comparison but at the point of creation.

**Review by shihchengli — APPROVED** [review]
All looks good! Thanks!
